File: src/SoftTouch.Parsing.Core/SDSL/Scanner.cs
namespace SoftTouch.Parsing.SDSL;

public struct Scanner
{

    readonly int start = 0;
    // public string Code { get; } = code;
    public readonly ReadOnlySpan<char> Code => Memory.Span;
    public ReadOnlyMemory<char> Memory { get; internal set; }
    public int Position { get; set; } = 0;

    public readonly int Line => Code[..Position].Count('\n') + 1;
    public readonly int Column => Position - Code[..Position].LastIndexOf('\n') + 1;


    public readonly int End => Code.Length;
    public readonly bool IsEof => Position >= End;

    public Scanner(string code)
    {
        Memory = code.AsMemory();
    }

    public Scanner(ReadOnlyMemory<char> code)
    {
        Memory = code;
    }


    public int ReadChar()
    {
        var pos = Position;
        if (pos < End)
        {
            Position = pos + 1;
            return Code[pos];
        }
        return -1;
    }

    public readonly int Peek()
    {
        var pos = Position;
        return pos < End ? Code[pos] : -1;
    }
    public readonly ReadOnlySpan<char> Peek(int size)
        => Position < End ? Slice(Position, size) : [];

    public int Advance(int length)
    {
        var pos = Position;
        var newPos = pos + length;
        if (newPos <= End)
        {
            Position = newPos;
            return pos;
        }
        return -1;
    }

    public readonly bool ReadString(string matchString, bool caseSensitive)
    {
        var index = Position;
        var Endstring = index + matchString.Length;
        if (Endstring <= End)
        {
            if (caseSensitive)
            {
                for (int i = 0; i < matchString.Length; i++)
                {
                    if (Code[index++] != matchString[i])
                        return false;
                }
                return true;
            }
            else
            {
                for (int i = 0; i < matchString.Length; i++)
                {
                    if (char.ToLowerInvariant(Code[index++]) != char.ToLowerInvariant(matchString[i]))
                        return false;
                }
                return true;
            }
        }
        return false;
    }

    public readonly ReadOnlySpan<char> Slice(int index, int length)
    {
        if (index < End)
        {
            length = Math.Min(index + length, End) - index;
            var slice = Code.Slice(index, length);
            return slice;
        }
        return [];
    }

    public readonly int LineAtIndex(int index)
    {
        int lineCount = 0;
        var max = Math.Min(End, index);
        for (int i = start; i < max; i++)
        {
            if (Code[i] == '\n')
                lineCount++;
        }
        return lineCount + 1;
    }

    public readonly TextLocation GetLocation(int position, int length)
    {
        return new(Line, Column, Memory.Slice(position, length));
    }
}

File: src/SoftTouch.Parsing.Core.FSharp/Scanner.fs
module Scanning
open System
open AST

let rec matchSensitiveChar (i : int32) (index : int32) (span : inref<ReadOnlySpan<char>>) (matchString : string): int32 =
    if span[index + i] = matchString.[i] then
        matchSensitiveChar (i + 1) index &span matchString
    else
        i

let rec matchChar (i : int32) (index : int32) (span : inref<ReadOnlySpan<char>>) (matchString : string): int32 =
    if Char.ToLowerInvariant(span[index + i]) <> Char.ToLowerInvariant(matchString.[i]) then
        matchChar (i + 1) index &span matchString
    else
        i

type Scanner =
    struct
        val Code : string
        val mutable Position : int32

        new(code : string) = { Code = code; Position = 0 }
        member this.Span : ReadOnlySpan<char> =
            this.Code.AsSpan()
        member this.Memory : ReadOnlyMemory<char> =
            this.Code.AsMemory()
        member this.IsEOL : bool =
            this.Position > this.Code.Length
        member this.End : int32 =
            this.Code.Length
        

        member this.Peek() : int32= 
            if this.Position + 1 < this.End && int this.Span[this.Position + 1] > 0 then  
                int this.Span[this.Position + 1]
            else 
                -1

        member this.Advance (length : int32) =
            let pos = this.Position;
            let newPos = pos + length;
            let mutable result = false
            if newPos <= this.End then
                this.Position <- newPos
                result <- true
            result

        member this.ReadString (matchString : string) (caseSensitive : bool) : bool =
            let endstring = this.Position + matchString.Length
            let mutable result = true

            if endstring <= this.End then
                let span = this.Span
                if caseSensitive then
                    if (matchSensitiveChar 0 this.Position &span matchString) < matchString.Length then
                        result <- false
                else
                    if (matchChar 0 this.Position &span matchString) < matchString.Length then
                        result <- false

            result
        
        member this.Slice position length = 
            this.Span.Slice(position, length)
        member this.BackSlice position = 
            this.Span.Slice(position, this.Position - position + 1)
        
        member this.GetLocation position length = 
            new TextLocation(position, length, this.Memory.Slice(position, length))
        member this.GetBackLocation position = 
            new TextLocation(position, this.Position - position, this.Memory.Slice(position, this.Position - position + 1))
    end